src/de/packets.rs

```diff
@@ -0,0 +1,421 @@
+    PingResp,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Reason<'a> {
+    #[serde(borrow)]
+    reason: Option<ReasonData<'a>>,
+}
+
+impl<'a> Reason<'a> {
+    pub fn code(&self) -> u8 {
+        self.reason.as_ref().map(|data| data.code).unwrap_or(0)
+    }
+
+    pub fn properties(&self) -> &'_ [Property<'a>] {
+        match &self.reason {
+            Some(ReasonData { properties, .. }) => properties,
+            _ => &[],
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ReasonData<'a> {
+    /// Reason code
+    pub code: u8,
+
+    /// The properties transmitted with the publish data.
+    #[serde(borrow)]
+    pub properties: Vec<Property<'a>, 8>,
+}
+
+struct ControlPacketVisitor;
+
+impl<'de> serde::de::Visitor<'de> for ControlPacketVisitor {
+    type Value = ReceivedPacket<'de>;
+
+    fn expecting(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
+        write!(formatter, "MQTT Control Packet")
+    }
+
+    fn visit_seq<A: serde::de::SeqAccess<'de>>(self, mut seq: A) -> Result<Self::Value, A::Error> {
+        use serde::de::Error;
+
+        let fixed_header: u8 = seq
+            .next_element()?
+            .ok_or_else(|| A::Error::custom("No header"))?;
+        let _length: Varint = seq
+            .next_element()?
+            .ok_or_else(|| A::Error::custom("No length"))?;
+        let packet_type = MessageType::try_from(fixed_header.get_bits(4..=7))
+            .map_err(|_| A::Error::custom("Invalid MQTT control packet type"))?;
+
+        let packet = match packet_type {
+            MessageType::ConnAck => ReceivedPacket::ConnAck(
+                seq.next_element()?
+                    .ok_or_else(|| A::Error::custom("No ConnAck"))?,
+            ),
+            MessageType::Publish => ReceivedPacket::Publish(
+                seq.next_element()?
+                    .ok_or_else(|| A::Error::custom("No Publish"))?,
+            ),
+            MessageType::PubAck => ReceivedPacket::PubAck(
+                seq.next_element()?
+                    .ok_or_else(|| A::Error::custom("No PubAck"))?,
+            ),
+            MessageType::SubAck => ReceivedPacket::SubAck(
+                seq.next_element()?
+                    .ok_or_else(|| A::Error::custom("No SubAck"))?,
+            ),
+            MessageType::PingResp => ReceivedPacket::PingResp,
+            MessageType::PubRec => ReceivedPacket::PubRec(
+                seq.next_element()?
+                    .ok_or_else(|| A::Error::custom("No PubRec"))?,
+            ),
+            MessageType::PubComp => ReceivedPacket::PubComp(
+                seq.next_element()?
+                    .ok_or_else(|| A::Error::custom("No PubComp"))?,
+            ),
+            MessageType::Disconnect => ReceivedPacket::Disconnect(
+                seq.next_element()?
+                    .ok_or_else(|| A::Error::custom("No Disconnect"))?,
+            ),
+            _ => return Err(A::Error::custom("Unsupported message type")),
```

Comment:
SImilar problem to that with properties. Can we change this mapping to a macro or some sort?